lib/ansible/modules/crypto/openssl_certificate.py

```diff
@@ -0,0 +1,771 @@
+
+        file_args = module.load_file_common_arguments(module.params)
+        if module.set_fs_attributes_if_different(file_args, False):
+            self.changed = True
+
+    def dump(self):
+
+        result = {
+            'changed': self.changed,
+            'filename': self.path,
+            'privatekey': self.privatekey_path,
+            'accountkey': self.accountkey_path,
+            'csr': self.csr_path,
+        }
+
+        return result
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            state=dict(default='present', choices=['present', 'absent'], type='str'),
+            path=dict(required=True, type='path'),
+            provider=dict(choices=['selfsigned', 'assertonly', 'acme'], type='str'),
+            force=dict(default=False, type='bool'),
+            csr_path=dict(type='path'),
+
+            # General properties of a certificate
+            privatekey_path=dict(type='path'),
+            privatekey_passphrase=dict(type='path', no_log=True),
+            signature_algorithms=dict(type='list'),
+            subject=dict(type='dict'),
+            issuer=dict(type='dict'),
+            has_expired=dict(default=False, type='bool'),
+            version=dict(type='int'),
+            keyUsage=dict(type='list'),
+            keyUsage_strict=dict(default=False, type='bool'),
+            extendedKeyUsage=dict(aliases=['extKeyUsage'], type='list'),
+            extendedKeyUsage_strict=dict(aliases=['extKeyUsage_strict'], default=False, type='bool'),
+            subjectAltName=dict(type='list'),
+            subjectAltName_strict=dict(default=False, type='bool'),
+            notBefore=dict(type='str'),
+            notAfter=dict(type='str'),
+            valid_at=dict(type='str'),
+            invalid_at=dict(type='str'),
+            valid_in=dict(type='int'),
+
+            # provider: selfsigned
+            selfsigned_digest=dict(default='sha256', type='str'),
+            selfsigned_notBefore=dict(type='str'),
+            selfsigned_notAfter=dict(type='str'),
+
+            # provider: acme
+            acme_accountkey_path=dict(type='path'),
+            acme_challenge_path=dict(type='path'),
+        ),
+        supports_check_mode=True,
+        add_file_common_args=True,
+    )
+
+    if not pyopenssl_found:
+        module.fail_json(msg='The python pyOpenSSL library is required')
+
+    base_dir = os.path.dirname(module.params['path'])
+    if not os.path.isdir(base_dir):
+        module.fail_json(
+            name=base_dir,
+            msg='The directory %s does not exist or the file is not a directory' % base_dir
+        )
+
+    provider = module.params['provider']
+
+    if provider == 'selfsigned':
+        certificate = SelfSignedCertificate(module)
+    elif provider == 'acme':
+        certificate = AcmeCertificate(module)
+    else:
+        certificate = AssertOnlyCertificate(module)
+
+    if module.params['state'] == 'present':
+
+        if module.check_mode:
```

Comment:
This is correct, because the only purpose of the check_mode test is to know if a resource will end up changed or not, and assert always return `changed = False`. So idiomatically, it wouldn't make sense to run the test in `assertonly()` for the check mode IMO